### PR TITLE
resolve source control dependencies via the registry when available

### PR DESCRIPTION
**Pull request by tomerd**
motivation: support mixed dependencies graphs where some dependencies come from source control while other from registry, and encourage the transition to the registry

changes:
* introduce new CLI flag to determine the source control to registry transformation mode (disabled, identity only, swizzle)
* wrap manifest loading with a transformer that translates source control dependencies to registry ones when the dependency is found in the registry, based on the mode
* add a bunch of tests

### CODE REVIEW

File: Sources/Workspace/Workspace.swift [hunk 1]
```diff
@@ -4185,6 +4189,169 @@ extension Workspace.Location {
     }
 }
 
+#warning("tomer: experiment")
+extension Workspace {
+    fileprivate struct RegistryAwareManifestLoader: ManifestLoaderProtocol {
```
Comment (tomerd) [inline Sources/Workspace/Workspace.swift:4194]: this is the key change. it basically translates source control dependencies to registries ones when possible.

this is still in wip state but the concept is clear

File: Sources/Workspace/Workspace.swift [hunk 2]
```diff
@@ -4185,6 +4189,219 @@ extension Workspace.Location {
...
+                        // an implementation mistake
+                        throw InternalError("unexpected non-source-control dependency")
+                    }
+                    switch settings.requirement {
+                    case .exact(let version):
+                        modifiedDependency = .registry(identity: registryIdentity, requirement: .exact(version), productFilter: settings.productFilter)
+                    case .range(let range):
+                        modifiedDependency = .registry(identity: registryIdentity, requirement: .range(range), productFilter: settings.productFilter)
+                    case .revision:
+                        // FIXME: warn?
```
Comment (yim-lee) [inline Sources/Workspace/Workspace.swift:4321]: Deprecation warning maybe? Though feels more like it should be "info" during the initial launch.

File: Sources/Commands/Options.swift [hunk 3]
```diff
@@ -267,38 +267,22 @@ public struct SwiftToolOptions: ParsableArguments {
...
-    var indexStoreMode: StoreMode = .autoIndexStore
-
     /// Whether to enable generation of `.swiftinterface`s alongside `.swiftmodule`s.
     @Flag(name: .customLong("enable-parseable-module-interfaces"))
     var shouldEnableParseableModuleInterfaces: Bool = false
 
-    /// Write dependency resolver trace to a file.
-    @Flag(name: .customLong("trace-resolver"), help: .hidden)
-    var _deprecated_enableResolverTrace: Bool = false
-
```
Comment (tomerd) [inline Sources/Commands/Options.swift:301]: small unrelated cleanup (moved code around)

File: Sources/Commands/Options.swift [hunk 4]
```diff
@@ -381,6 +365,26 @@ public struct SwiftToolOptions: ParsableArguments {
         help: "Disable/enable dead code stripping by the linker")
     var linkerDeadStrip: Bool = true
 
+    @Flag(help: "Define automatic transformation of source control based dependencies to registry based ones")
+    var sourceControlToRegistryDependencyTransformation: SourceControlToRegistryDependencyTransformation = .disabled
```
Comment (tomerd) [inline Sources/Commands/Options.swift:369]: the new flag, it a mouth full!

File: Sources/Commands/SwiftTool.swift [hunk 5]
```diff
@@ -1114,6 +1115,13 @@ extension ObservabilityMetadata {
     }
 }
 
+extension Workspace.ManagedDependency {
+    fileprivate var isEdited: Bool {
+        if case .edited = self.state { return true }
+        return false
+    }
+}
```
Comment (tomerd) [inline Sources/Commands/SwiftTool.swift:1123]: the code that moved around

File: Sources/Workspace/Workspace.swift [hunk 6]
```diff
@@ -4185,6 +4189,219 @@ extension Workspace.Location {
...
+        }
+
+        private func mapRegistryIdentity(
+            url: URL,
+            observabilityScope: ObservabilityScope,
+            callbackQueue: DispatchQueue,
+            completion: @escaping (Result<PackageIdentity?, Error>) -> Void
+        ) {
+            self.registryClient.lookupIdentities(url: url, observabilityScope: observabilityScope, callbackQueue: callbackQueue) { result in
+                // FIXME: returns first result... need to consider how to address multuiple ones
```
[the remark this answers is not in the document]
Comment (tomerd) [reply Sources/Workspace/Workspace.swift:4445]: yes, will address in follow up PR

File: Sources/Workspace/Workspace.swift [hunk 7]
```diff
@@ -4185,6 +4194,290 @@ extension Workspace.Location {
...
+                }
+            }
+        }
+
+        private func transformManifest(
+            manifest: Manifest,
+            transformations: [PackageDependency: PackageIdentity],
+            transformationMode: TransformationMode,
+            observabilityScope: ObservabilityScope
+        ) throws -> Manifest {
```
Comment (tomerd) [inline Sources/Workspace/Workspace.swift:4321]: this performs the swizzling and the heart of this feature

File: Sources/Workspace/Workspace.swift [hunk 8]
```diff
@@ -4185,6 +4189,219 @@ extension Workspace.Location {
...
+        func purgeCache() throws {
+            try self.underlying.purgeCache()
+        }
+
+        // transforms source control dependencies to registry based ones when appropriate
+        // FIXME:
+        // 1. what to do with revision / branch based deps
+        // 2. add flag to turn this look up off?
+        // 3. what about a mixed situation when some versions on the registry but some on source control? do we need a second lookup that is version based?
+        // 4. check if registry is even configured, or not configured with a default
```
[the remark this answers is not in the document]
Comment (tomerd) [reply Sources/Workspace/Workspace.swift]: yes

File: Sources/Workspace/Workspace.swift (hunk 2 above)
Comment (tomerd) [reply Sources/Workspace/Workspace.swift]: found a solution in which we only replace the identity instead of full swizzle. works well!

File: Sources/Commands/Options.swift [hunk 9]
```diff
@@ -381,6 +365,26 @@ public struct SwiftToolOptions: ParsableArguments {
...
+        case swizzle
+
+        static func name(for value: Self) -> NameSpecification {
+            switch value {
+            case .disabled:
+                return .customLong("disable-scm-to-regsitry-lookup")
+            case .identity:
+                return .customLong("use-registry-identity-for-scm")
+            case .swizzle:
+                return .customLong("raplace-scm-with-registry")
```
Comment (abertelrud) [inline Sources/Commands/Options.swift:383]:
```suggestion
                return .customLong("replace-scm-with-registry")
```

Comment (abertelrud) [reply Sources/Commands/Options.swift:383]: Is there a way to add help strings for each of the modes?  The longer names are useful but the distinction between the last two seems a bit nuanced.

File: Sources/Commands/Options.swift [hunk 10]
```diff
@@ -381,6 +365,26 @@ public struct SwiftToolOptions: ParsableArguments {
...
+
+    enum SourceControlToRegistryDependencyTransformation: EnumerableFlag {
+        case disabled
+        case identity
+        case swizzle
+
+        static func name(for value: Self) -> NameSpecification {
+            switch value {
+            case .disabled:
+                return .customLong("disable-scm-to-regsitry-lookup")
```
Comment (abertelrud) [inline Sources/Commands/Options.swift:379]:
```suggestion
                return .customLong("disable-scm-to-registry-lookup")
```

File: Sources/PackageModel/Manifest/PackageDependencyDescription.swift [hunk 11]
```diff
@@ -12,19 +12,19 @@ import Foundation
 import TSCBasic
 
 /// Represents a package dependency.
-public enum PackageDependency: Equatable {
+public enum PackageDependency: Equatable, Hashable {
```
Comment (abertelrud) [inline Sources/PackageModel/Manifest/PackageDependencyDescription.swift:15]: Hashable implies Equatable, but it makes sense to me as a style thing to list both.  I think I've been doing just Hashable, should we switch to consistently listing both in this code base going forward?  I kind of prefer it for clarity now that I see it.

File: Sources/SPMTestSupport/MockPackage.swift [hunk 12]
```diff
@@ -73,6 +73,7 @@ public struct MockPackage {
         name: String,
         platforms: [PlatformDescription] = [],
         identity: String,
+        alternativeURLs: [String]? = .none,
```
Comment (abertelrud) [inline Sources/SPMTestSupport/MockPackage.swift:76]: Is there a semantic distinction between `[]` and `.none` here?  i.e. should it be maybe be non-optional?

File: Sources/SPMTestSupport/MockRegistry.swift [hunk 13]
```diff
@@ -94,13 +107,21 @@ public class MockRegistry {
     private func handleRequest(request: HTTPClient.Request) throws -> HTTPClient.Response {
         let routeComponents = request.url.absoluteString.dropFirst(Self.mockRegistryURL.absoluteString.count + 1).split(separator: "/")
         switch routeComponents.count {
+        case _ where routeComponents[0].hasPrefix("identifiers?url="):
+            guard let query = request.url.query else {
+                throw StringError("invalid url: \(request.url)")
+            }
+            guard let sourceControlURL = URL(string: String(query.dropFirst(4))) else {
```
Comment (abertelrud) [inline Sources/SPMTestSupport/MockRegistry.swift:114]: Are these four characters always the `url=` mentioned above, and therefore no additional check is needed for what they are?

File: Sources/SPMTestSupport/MockRegistry.swift [hunk 14]
```diff
@@ -114,21 +135,33 @@ public class MockRegistry {
...
             }
         )
 
+        /*
+         <https://github.com/mona/LinkedList>; rel="canonical",
+         <ssh://git@github.com:mona/LinkedList.git>; rel="alternate"
+         */
+        let sourceControlURLs = self.packagesSourceControlURLs[packageIdentity]
+        let links = sourceControlURLs?.map { url in
+            "<\(url.absoluteString)>; rel=alternate"
```
Comment (abertelrud) [inline Sources/SPMTestSupport/MockRegistry.swift:156]: Does this `absoluteString` result need escaping if it contains characters that have special meanings?

File: Sources/Workspace/Workspace.swift [hunk 15]
```diff
@@ -4185,6 +4194,290 @@ extension Workspace.Location {
...
+        ) throws -> Manifest {
+            var targetDependencyPackageNameTransformations = [String: String]()
+
+            var modifiedDependencies = [PackageDependency]()
+            for dependency in manifest.dependencies {
+                var modifiedDependency = dependency
+                if let registryIdentity = transformations[dependency] {
+                    guard case .sourceControl(let settings) = dependency, case .remote = settings.location else {
+                        // an implementation mistake
+                        throw InternalError("unexpected non-source-control dependency")
```
Comment (abertelrud) [inline Sources/Workspace/Workspace.swift:4330]: Should the error contain the dependencies to make it easier to debug if this is every thrown?

**Review by abertelrud — APPROVED** [review]

File: Sources/Commands/Options.swift (hunk 9 above)
Comment (tomerd) [reply Sources/Commands/Options.swift:383]: good idea. will do

File: Sources/PackageModel/Manifest/PackageDependencyDescription.swift (hunk 11 above)
Comment (tomerd) [reply Sources/PackageModel/Manifest/PackageDependencyDescription.swift:15]: its a style thing. we have a mixed bag in the code base today. probably topic for a cleanup PR

File: Sources/SPMTestSupport/MockPackage.swift (hunk 12 above)
Comment (tomerd) [reply Sources/SPMTestSupport/MockPackage.swift:76]: its a style thing, I guess (not sure about memory implication, empty array maybe also memory efficient). I personally prefer to use optionals when the data is semantically optional

File: Sources/SPMTestSupport/MockRegistry.swift (hunk 13 above)
Comment (tomerd) [reply Sources/SPMTestSupport/MockRegistry.swift:114]: this is in the testing infrastructure so should be safe enough

File: Sources/SPMTestSupport/MockRegistry.swift (hunk 14 above)
Comment (tomerd) [reply Sources/SPMTestSupport/MockRegistry.swift:156]: this is in the testing infrastructure so should be safe enough

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test
